src/oracle/connection/stmt.rs

```diff
@@ -200,77 +177,79 @@ impl Statement {
         check
     }
 
-    pub fn run(&mut self, auto_commit: bool) -> QueryResult<()> {
+    pub fn run(&mut self, auto_commit: bool, metadata: &[OCIDataType]) -> QueryResult<()> {
         let iters = if self.is_select { 0 } else { 1 };
+        let mut octx = Vec::new();
+
         if self.is_returning {
-            self.bind_index += 1;
-        }
-        let mut octx = BindContext::new(self.connection.env.error_handle);
-        if self.is_returning {
-            let tpe = OCIDataType::Char;
-            let mut bndp = ptr::null_mut() as *mut ffi::OCIBind;
-
-            unsafe {
-                // read https://docs.oracle.com/database/121/LNOCI/oci16rel003.htm#LNOCI153
-                // read it again, then you will understand why the parameters are set like that
-                // make sure to read it again
-                // otherwise you may enter the ORA-03106: fatal two-task communication protocol error-hell
-                let status = ffi::OCIBindByPos(
-                    self.inner_statement,
-                    &mut bndp,
-                    self.connection.env.error_handle,
-                    self.bind_index,
-                    ptr::null_mut(),
-                    NUM_ELEMENTS as i32,
-                    tpe.to_raw() as u16,
-                    ptr::null_mut(),
-                    ptr::null_mut(),
-                    ptr::null_mut(),
-                    0,
-                    ptr::null_mut(),
-                    ffi::OCI_DATA_AT_EXEC,
-                );
+            octx.reserve_exact(metadata.len());
+            for tpe in metadata {
+                self.bind_index +=1;
+                octx.push(BindContext::new(self.connection.env.error_handle));
+                let octx = octx.last_mut().expect("We pushed it above");
```

Comment:
I'd prefer a different variable name. Maybe rustc can handle it, but my brain not. Sorry.